diff --git a/core/main.mk b/core/main.mk
index c63c6dfde9..10cb30ff55 100644
--- a/core/main.mk
+++ b/core/main.mk
@@ -748,17 +748,17 @@ ifneq (,$(filter $(HOST_OS),darwin))
   check_missing_required_modules :=
 endif # HOST_OS == darwin
 
-ifeq (true,$(check_missing_required_modules))
-ifneq (,$(_nonexistent_required))
-  $(warning Missing required dependencies:)
-  $(foreach r_i,$(_nonexistent_required), \
-    $(eval r := $(subst $(comma),$(space),$(r_i))) \
-    $(info $(word 1,$(r)) module $(word 2,$(r)) requires non-existent $(word 3,$(r)) module: $(word 4,$(r))) \
-  )
-  $(warning Set BUILD_BROKEN_MISSING_REQUIRED_MODULES := true to bypass this check if this is intentional)
-  $(error Build failed)
-endif # _nonexistent_required != empty
-endif # check_missing_required_modules == true
+# ifeq (true,$(check_missing_required_modules))
+# ifneq (,$(_nonexistent_required))
+#   $(warning Missing required dependencies:)
+#   $(foreach r_i,$(_nonexistent_required), \
+#     $(eval r := $(subst $(comma),$(space),$(r_i))) \
+#     $(info $(word 1,$(r)) module $(word 2,$(r)) requires non-existent $(word 3,$(r)) module: $(word 4,$(r))) \
+#   )
+#   $(warning Set BUILD_BROKEN_MISSING_REQUIRED_MODULES := true to bypass this check if this is intentional)
+#   $(error Build failed)
+# endif # _nonexistent_required != empty
+# endif # check_missing_required_modules == true
 
 define add-required-deps
 $(1): | $(2)
diff --git a/target/product/gsi/33.txt b/target/product/gsi/33.txt
index 03a143d02c..060be5d09d 100644
--- a/target/product/gsi/33.txt
+++ b/target/product/gsi/33.txt
@@ -180,6 +180,7 @@ VNDK-core: libwifi-system-iface.so
 VNDK-core: libxml2.so
 VNDK-core: libyuv.so
 VNDK-core: libziparchive.so
+VNDK-core: android.hardware.emvco-V1-ndk.so
 VNDK-private: libbacktrace.so
 VNDK-private: libblas.so
 VNDK-private: libcompiler_rt.so
diff --git a/target/product/gsi/current.txt b/target/product/gsi/current.txt
index 03a143d02c..060be5d09d 100644
--- a/target/product/gsi/current.txt
+++ b/target/product/gsi/current.txt
@@ -180,6 +180,7 @@ VNDK-core: libwifi-system-iface.so
 VNDK-core: libxml2.so
 VNDK-core: libyuv.so
 VNDK-core: libziparchive.so
+VNDK-core: android.hardware.emvco-V1-ndk.so
 VNDK-private: libbacktrace.so
 VNDK-private: libblas.so
 VNDK-private: libcompiler_rt.so
